File: src/services/UsageRefreshService.ts
// UsageRefreshService —— 架构 §4.4 / §5。
// 唯一允许执行刷新流程的服务。负责：节流、并发锁、连续失败退避、认证/解析失败停止自动刷新。
// 平台无关纯逻辑：网络/存储/UI 更新通过依赖注入解耦，移植 ArkTS 时只换实现、编排逻辑不动。

import type { CodingPlanUsage, UsageErrorCode, UsageStatus } from '../domain/types.ts';
import type { UsageCache } from './UsageCache.ts';
import type { MappedError } from './ErrorMapper.ts';

export type RefreshReason = 'manual' | 'foreground' | 'background';

export interface UsageProviderLike {
  fetchUsage(): Promise<CodingPlanUsage>;
}

export interface RefreshSettings {
  /** 后台自动刷新开关（PRD §6.4，默认关） */
  autoRefreshEnabled: boolean;
}

export interface RefreshServiceOptions {
  provider: UsageProviderLike;
  cache: UsageCache;
  /** 注入时钟，便于测试 */
  now: () => number;
  /** 每次刷新结果（成功或失败快照）都触发，供 UI / Form 更新 */
  onRefresh?: (usage: CodingPlanUsage) => void;
  /** 读取用户设置（后台刷新开关等） */
  getSettings?: () => RefreshSettings;
}

// 时长常量（架构 §5.1 / §5.2 / PRD §8.1）
const THROTTLE_MS = 10_000;                    // 手动节流 10s
const FOREGROUND_INTERVAL_MS = 15 * 60_000;    // 前台 15min
const BACKGROUND_INTERVAL_MS = 60 * 60_000;    // 后台 60min
const FAILURE_THRESHOLD = 3;                   // 连续失败 3 次
const AUTO_PAUSE_MS = 6 * 60 * 60_000;         // 暂停自动刷新 6h

/** 这些错误码意味着自动刷新必须停止（仅允许手动），直到用户重新配置/升级 */
const STOP_CODES: ReadonlySet<UsageErrorCode> = new Set(['AUTH', 'NO_PLAN', 'UPSTREAM_CHANGED']);

export class UsageRefreshService {
  private lastFetchStartedAt = 0;
  private lastSuccessAt = 0;
  private consecutiveFailures = 0;
  private autoRefreshPausedUntil = 0;
  private stoppedFor: UsageErrorCode | null = null;
  private lastUsage: CodingPlanUsage | null = null;
  private inFlight: Promise<CodingPlanUsage> | null = null;

  private readonly opts: RefreshServiceOptions;

  constructor(opts: RefreshServiceOptions) {
    this.opts = opts;
  }

  /** 是否允许发起刷新（不实际发起）。供调用方决定是否触发后台任务。 */
  canRefresh(reason: RefreshReason): boolean {
    const now = this.opts.now();
    if (reason === 'manual') {
      // 手动：仅受 10s 节流约束；从未请求过时直接允许（lastFetchStartedAt===0 为哨兵）
      return this.lastFetchStartedAt === 0 || now - this.lastFetchStartedAt >= THROTTLE_MS;
    }
    if (this.stoppedFor !== null) return false;
    if (now < this.autoRefreshPausedUntil) return false;
    if (reason === 'background' && !(this.opts.getSettings?.().autoRefreshEnabled ?? false)) return false;
    const interval = reason === 'foreground' ? FOREGROUND_INTERVAL_MS : BACKGROUND_INTERVAL_MS;
    // 从未成功过时直接允许（建立首次数据）
    return this.lastSuccessAt === 0 || now - this.lastSuccessAt >= interval;
  }

  /**
   * 发起刷新。并发触发复用进行中的 Promise（架构 §4.4）。
   * 总是 resolve 一个 CodingPlanUsage：成功 status=ok；失败保留上次缓存标 stale，或无缓存返回 error 态。
   */
  async refresh(reason: RefreshReason): Promise<CodingPlanUsage> {
    if (this.inFlight) return this.inFlight;
    if (!this.canRefresh(reason)) return this.currentSnapshot();
    const p = this.doRefresh();
    this.inFlight = p;
    try {
      return await p;
    } finally {
      this.inFlight = null;
    }
  }

  /** 当前快照（内存中的最后结果），供不刷新时展示 */
  currentSnapshot(): CodingPlanUsage {
    if (this.lastUsage) return this.lastUsage;
    return {
      session: { usedPercent: 0, remainingPercent: 0 },
      weekly: { usedPercent: 0, remainingPercent: 0 },
      updatedAt: 0,
      source: 'direct',
      status: 'unconfigured',
    };
  }

  /** 冷启动：从持久化缓存加载到内存（架构 §8.3） */
  async hydrateFromCache(): Promise<CodingPlanUsage | null> {
    const cached = await this.opts.cache.load();
    if (cached) {
      this.lastUsage = cached.usage;
      this.lastSuccessAt = cached.usage.updatedAt;
    }
    return cached ? cached.usage : null;
  }

  private async doRefresh(): Promise<CodingPlanUsage> {
    this.lastFetchStartedAt = this.opts.now();
    try {
      const usage = await this.opts.provider.fetchUsage();
      // 成功：覆盖缓存、重置失败计数与暂停/停止状态
      this.lastUsage = usage;
      this.lastSuccessAt = this.opts.now();
      this.consecutiveFailures = 0;
      this.autoRefreshPausedUntil = 0;
      this.stoppedFor = null;
      await this.opts.cache.save(usage);
      this.opts.onRefresh?.(usage);
      return usage;
    } catch (e) {
      const mapped: MappedError = (e as { mapped?: MappedError })?.mapped ?? {
        code: 'UNKNOWN',
        message: '暂时无法获取用量',
      };
      const snapshot = this.applyFailure(mapped);
      this.opts.onRefresh?.(snapshot);
      return snapshot;
    }
  }

  private applyFailure(mapped: MappedError): CodingPlanUsage {
    const { code, message } = mapped;
    if (STOP_CODES.has(code)) {
      // 认证/无套餐/解析失败：停止自动刷新（架构 §5.2）
      this.stoppedFor = code;
    } else {
      // 网络/限流/未知：计入连续失败，达阈值则暂停 6h
      this.consecutiveFailures += 1;
      if (this.consecutiveFailures >= FAILURE_THRESHOLD) {
        this.autoRefreshPausedUntil = this.opts.now() + AUTO_PAUSE_MS;
      }
    }
    const status: UsageStatus = this.lastUsage ? 'stale' : 'error';
    if (this.lastUsage) {
      // 保留最后有效数据，仅改状态与错误信息
      return { ...this.lastUsage, status, errorCode: code, errorMessage: message };
    }
    return {
      session: { usedPercent: 0, remainingPercent: 0 },
      weekly: { usedPercent: 0, remainingPercent: 0 },
      updatedAt: this.opts.now(),
      source: 'direct',
      status: 'error',
      errorCode: code,
      errorMessage: message,
    };
  }
}

File: android/app/src/main/java/com/example/myapplication/services/UsageRefreshService.kt
package com.example.myapplication.services

import com.example.myapplication.domain.CodingPlanUsage
import com.example.myapplication.domain.ServiceProviderInfo
import com.example.myapplication.domain.UsageErrorCode
import com.example.myapplication.domain.UsageSource
import com.example.myapplication.domain.UsageStatus
import com.example.myapplication.domain.UsageWindow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** 刷新编排（架构 §4.4/§5）：节流、并发锁、连续失败退避、认证/解析失败停止自动刷新。 */
class UsageRefreshService(
    private val provider: UsageProvider,
    private val cacheStorage: CacheStorage,
    private val now: () -> Long = { System.currentTimeMillis() },
    private val onRefresh: ((CodingPlanUsage) -> Unit)? = null,
    private val getSettings: () -> RefreshSettings = { RefreshSettings(false) }
) {
    data class RefreshSettings(val autoRefreshEnabled: Boolean)
    enum class Reason { MANUAL, FOREGROUND, BACKGROUND }

    private val mutex = Mutex()
    private var lastFetchStartedAt = 0L
    private var lastSuccessAt = 0L
    private var consecutiveFailures = 0
    private var autoRefreshPausedUntil = 0L
    private var stoppedFor: UsageErrorCode? = null
    private var lastUsage: CodingPlanUsage? = null

    fun canRefresh(reason: Reason): Boolean {
        val now = now()
        if (reason == Reason.MANUAL) {
            return lastFetchStartedAt == 0L || now - lastFetchStartedAt >= THROTTLE_MS
        }
        if (stoppedFor != null) return false
        if (now < autoRefreshPausedUntil) return false
        if (reason == Reason.BACKGROUND && !getSettings().autoRefreshEnabled) return false
        val interval = if (reason == Reason.FOREGROUND) FOREGROUND_MS else BACKGROUND_MS
        return lastSuccessAt == 0L || now - lastSuccessAt >= interval
    }

    /** 成功返回 ok 用量；失败保留缓存标 stale，或无缓存返回 error 态。并发由 Mutex 串行化。 */
    suspend fun refresh(reason: Reason): CodingPlanUsage = mutex.withLock {
        if (!canRefresh(reason)) return@withLock currentSnapshot()
        lastFetchStartedAt = now()
        try {
            val usage = provider.fetchUsage()
            lastUsage = usage
            lastSuccessAt = now()
            consecutiveFailures = 0
            autoRefreshPausedUntil = 0L
            stoppedFor = null
            UsageCache.save(cacheStorage, usage)
            onRefresh?.invoke(usage)
            usage
        } catch (e: UsageProviderException) {
            val snap = applyFailure(e.mapped)
            onRefresh?.invoke(snap)
            snap
        }
    }

    fun currentSnapshot(): CodingPlanUsage = lastUsage ?: CodingPlanUsage(
        session = UsageWindow(0, 100),
        weekly = UsageWindow(0, 100),
        updatedAt = 0L,
        source = UsageSource.DIRECT,
        providerLabel = ServiceProviderInfo.GLM_LABEL,
        status = UsageStatus.UNCONFIGURED
    )

    suspend fun hydrateFromCache(): CodingPlanUsage? {
        val cached = UsageCache.load(cacheStorage)
        if (cached != null) {
            lastUsage = cached
            lastSuccessAt = cached.updatedAt
        }
        return cached
    }

    private fun applyFailure(mapped: MappedError): CodingPlanUsage {
        val code = mapped.code
        if (code in STOP_CODES) {
            stoppedFor = code
        } else {
            consecutiveFailures++
            if (consecutiveFailures >= FAILURE_THRESHOLD) {
                autoRefreshPausedUntil = now() + AUTO_PAUSE_MS
            }
        }
        val status = if (lastUsage != null) UsageStatus.STALE else UsageStatus.ERROR
        return if (lastUsage != null) {
            lastUsage!!.copy(status = status, errorCode = code, errorMessage = mapped.message)
        } else {
            CodingPlanUsage(
                session = UsageWindow(0, 0),
                weekly = UsageWindow(0, 0),
                updatedAt = now(),
                source = UsageSource.DIRECT,
                providerLabel = ServiceProviderInfo.GLM_LABEL,
                status = status,
                errorCode = code,
                errorMessage = mapped.message
            )
        }
    }

    companion object {
        private const val THROTTLE_MS = 10_000L
        private const val FOREGROUND_MS = 15 * 60_000L
        private const val BACKGROUND_MS = 60 * 60_000L
        private const val FAILURE_THRESHOLD = 3
        private const val AUTO_PAUSE_MS = 6 * 60 * 60_000L
        private val STOP_CODES =
            setOf(UsageErrorCode.AUTH, UsageErrorCode.NO_PLAN, UsageErrorCode.UPSTREAM_CHANGED)
    }
}
